src/impl_methods.rs

```diff
@@ -441,33 +455,35 @@ where
 
     /// Slice the array in place without changing the number of dimensions.
     ///
-    /// Note that [`&SliceInfo`](struct.SliceInfo.html) (produced by the
-    /// [`s![]`](macro.s!.html) macro) will usually coerce into `&D::SliceArg`
-    /// automatically, but in some cases (e.g. if `D` is `IxDyn`), you may need
-    /// to call `.as_ref()`.
+    /// Note that `NewAxis` elements in `info` are ignored.
     ///
     /// See [*Slicing*](#slicing) for full documentation.
-    /// See also [`D::SliceArg`].
-    ///
-    /// [`D::SliceArg`]: trait.Dimension.html#associatedtype.SliceArg
     ///
     /// **Panics** if an index is out of bounds or step size is zero.<br>
-    /// (**Panics** if `D` is `IxDyn` and `indices` does not match the number of array axes.)
-    pub fn slice_collapse(&mut self, indices: &D::SliceArg) {
-        let indices: &[SliceOrIndex] = indices.as_ref();
-        assert_eq!(indices.len(), self.ndim());
-        indices
-            .iter()
-            .enumerate()
-            .for_each(|(axis, &slice_or_index)| match slice_or_index {
-                SliceOrIndex::Slice { start, end, step } => {
-                    self.slice_axis_inplace(Axis(axis), Slice { start, end, step })
+    /// (**Panics** if `D` is `IxDyn` and `info` does not match the number of array axes.)
+    pub fn slice_collapse<I>(&mut self, info: &I)
+    where
+        I: CanSlice<D>,
+    {
+        assert_eq!(
+            info.in_ndim(),
+            self.ndim(),
+            "The input dimension of `info` must match the array to be sliced.",
+        );
+        let mut axis = 0;
+        info.as_ref().iter().for_each(|&ax_info| match ax_info {
+                AxisSliceInfo::Slice { start, end, step } => {
+                    self.slice_axis_inplace(Axis(axis), Slice { start, end, step });
+                    axis += 1;
                 }
-                SliceOrIndex::Index(index) => {
+                AxisSliceInfo::Index(index) => {
                     let i_usize = abs_index(self.len_of(Axis(axis)), index);
-                    self.collapse_axis(Axis(axis), i_usize)
+                    self.collapse_axis(Axis(axis), i_usize);
+                    axis += 1;
                 }
+                AxisSliceInfo::NewAxis => {}
```

Comment:
What would happen if it was an error instead?

Reply:
That's probably better than just ignoring it silently. I've made that change to the PR. Other options include:

- Panic if `info` contains `NewAxis`.
- Prevent this at compile time. (Add another type parameter to `SliceInfo` to track whether or not it contains `NewAxis`, and add another trait like `CanSlice`, but specifically for `slice_collapse`. We'll have to add another type parameter to support ellipsis anyway, so maybe the additional complexity of this option wouldn't be too bad.)

Reply:
I've changed it to a panic, based on @bluss's comment below.